components/Sataserver/src/sataserver.c

```diff
@@ -0,0 +1,663 @@
+    error = sel4utils_make_iommu_dma_alloc(&vka, &vspace, &ioops.dma_manager, num_bdfs, iospace_caps);
+    assert(!error);
+    ioops.dma_manager.dma_alloc_fn = camkes_iommu_dma_alloc;
+    error = sel4platsupport_get_io_port_ops(&ioops.io_port_ops, &camkes_simple, &vka);
+    assert(!error);
+
+#ifdef CONFIG_AHCI_ENABLE
+    /* preallocate buffers */
+    int clb_buf_size = BUF_SIZE;
+    void *clb_buf = ps_dma_alloc(&ioops.dma_manager, clb_buf_size, 4, 0, PS_MEM_NORMAL);
+    assert(clb_buf);
+    uintptr_t phys = ps_dma_pin(&ioops.dma_manager, clb_buf, clb_buf_size);
+    assert(phys == (uintptr_t)clb_buf);
+
+    int ctba_buf_size = BUF_SIZE * 16;
+    void *ctba_buf = ps_dma_alloc(&ioops.dma_manager, ctba_buf_size, 4, 0, PS_MEM_NORMAL);
+    assert(ctba_buf);
+    uintptr_t phys2 = ps_dma_pin(&ioops.dma_manager, ctba_buf, clb_buf_size);
+    assert(phys2 == (uintptr_t)ctba_buf);
+
+    int fb_buf_size = BUF_SIZE;
+    void *fb_buf = ps_dma_alloc(&ioops.dma_manager, fb_buf_size, 4, 0, PS_MEM_NORMAL);
+    assert(fb_buf);
+    uintptr_t phys3 = ps_dma_pin(&ioops.dma_manager, fb_buf, fb_buf_size);
+    assert(phys3 == (uintptr_t)fb_buf);
+
+    int data_buf_size = BUF_SIZE;
+    void *data_buf = ps_dma_alloc(&ioops.dma_manager, data_buf_size, 4, 0, PS_MEM_NORMAL);
+    assert(data_buf);
+    uintptr_t phys4 = ps_dma_pin(&ioops.dma_manager, data_buf, data_buf_size);
+    assert(phys4 == (uintptr_t)data_buf);
+
+    /* Initialize the device */
+    ahci_intel_config_t ahci_config = (ahci_intel_config_t) {
+        .bar0      = (void *)ahcidriver,
+        .clb       = clb_buf,
+        .clb_size  = clb_buf_size,
+        .ctba      = ctba_buf,
+        .ctba_size = ctba_buf_size,
+        .fb        = fb_buf,
+        .fb_size   = fb_buf_size,
+        .data      = data_buf,
+        .data_size = data_buf_size
+    };
+
+    error = sata_init(&ioops, &ahci_config);
+    assert(!error);
+#else
+    error = sata_init(&ioops, NULL);
+    assert(!error);
+#endif
+
+
+    /* Get the SATA device partition tables, We assume drive 0 for now.
+     *    TBD: Handle different/dynamic drives
+     */
+    sata_get_partition_tables(drive, phys_partition_tables, g_part_data);
+
+    uint8_t partition;
+
+    /* Determine the number of clients and allocate memory for their structures */
+    num_clients = client_num_badges();
+    clients = calloc(num_clients, sizeof(client_t));
+
+    /* Get and validate client specific data for each client connected to the sataserver */
+    for (int client = 0; client < num_clients; client++) {
+
+        /* Get the information from the configuration */
+        clients[client].client_id = client_enumerate_badge(client);
+        clients[client].dataport = client_buf(clients[client].client_id);
+        client_get_partitions(clients[client].client_id, clients[client].partitions, &clients[client].num_partitions);
+        ZF_LOGE("SATASERVER: Client %u Partitions: ", clients[client].client_id);
```

Comment:
Can these log levels be changed to Info?

Reply:
Yep.